### PR TITLE
FM2-352: Add support for creating encounters via the FHIR API

### PR DESCRIPTION
**Pull request by Medhavi-16**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-352

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by Medhavi-16** [comment]
@ibacher I added support for adding encounters through FHIR API. PTAL.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java [hunk 1]
```diff
@@ -73,6 +70,57 @@ public Encounter get(@Nonnull String uuid) {
 		return result;
 	}
 	
+	@Override
+	public Encounter create(@Nonnull Encounter encounter) {
+		
+		if (encounter.getTypeFirstRep().getCodingFirstRep().getSystem() == null) {
+			throw new InvalidRequestException("Id type must be specified to update");
+		}
+		if (encounter.getTypeFirstRep().getCodingFirstRep().getSystem().equals(FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI)) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:79]: So, rather than relying on the *first* type coding (especially since we want to ignore other coding types), I'd refactor this into a utility method, something like:

```java
private enum OpenmrsEncounterType {
	ENCOUNTER,
	VISIT,
	AMBIGUOUS
}

// ...

private Optional<OpenmrsEncounterType> getOpenmrsEncounterType(Encounter encounter) {
	List<OpenmrsEncounterType> openmrsEncounterTypes = encounter.getType().stream().flatMap(it -> 
		it.getCoding().stream()
			.filter(coding -> FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem()) ||
				FhirConstants.VISIT_TYPE_SYSTEM_URI.equals(coding.getSystem()))
			.map(coding -> {
				if (FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem())) {
					return OpenmrsEncounterType.ENCOUNTER;
				} else {
					return OpenmrsEncounterType.VISIT;
				}
			})
	).distinct().collect(Collectors.toList());
	
	if (openmrsEncounterTypes.isEmpty()) {
		return Optional.empty();
	} else if (openmrsEncounterTypes.size() == 1) {
		return Optional.of(openmrsEncounterTypes.get(0));
	} else {
		return Optional.of(OpenmrsEncounterType.AMBIGUOUS);
	}
}
```

Now as we end up with three states:

1. We known the encounter type of the encounter and can process it
2. No encounter type was supplied and we can throw an exception with a message to that effect
3. An encounter invalidly specified that it is *both* an OpenMRS Encounter and an OpenMRS visit and we can send an appropriate error message for that.

Plus, we don't depend on the order of the types or do anything else in response to codings we don't care about.

File: api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImplTest.java [hunk 2]
```diff
@@ -102,7 +106,12 @@
 	
 	@Before
 	public void setUp() {
-		encounterService = new FhirEncounterServiceImpl();
+		encounterService = new FhirEncounterServiceImpl() {
+			
+			@Override
+			protected void validateObject(Encounter object) {
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImplTest.java:112]: Why do we need to stub this out?

File: api/src/main/java/org/openmrs/module/fhir2/FhirConstants.java [hunk 3]
```diff
@@ -50,7 +50,7 @@
 	        + "/condition-ver-status";
 	
 	public static final String CLINICAL_FINDINGS_SYSTEM_URI = HL7_FHIR_CODE_SYSTEM_PREFIX + "/clinical-findings";
-
+
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/FhirConstants.java:53]: Can we exclude this change from this PR? It prevents it from interfering with other PRs that might make the same change.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java [hunk 4]
```diff
@@ -58,9 +58,6 @@
 	
 	@Override
 	public Encounter get(@Nonnull String uuid) {
-		if (uuid == null) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:61]: Why did you remove this check? Note that `@Nonnull` is a hint to the programmer; it doesn't guarantee that the value here is never null (and we should have the same checks we do [in the base class](https://github.com/openmrs/openmrs-module-fhir2/blob/master/api/src/main/java/org/openmrs/module/fhir2/api/impl/BaseFhirService.java)).

**Review by ibacher — CHANGES_REQUESTED** [review]
@Medhavi-16 Thanks for this. It mostly looks good. I've added a few comments on some changes I'd like to see. In addition, it would be good to re-create the encounter related tests that were removed in [this commit](https://github.com/openmrs/openmrs-module-fhir2/commit/c6fc65a105b7a875cf9973966274c1880933c842#diff-d29233104de9d96ca6e8af497f647719400e78b850b206e4dddbbd5d74b22a76) or equivalents of them so we can be sure the changes actually work.

File: api/src/main/java/org/openmrs/module/fhir2/FhirConstants.java (hunk 3 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/FhirConstants.java:53]: @ibacher Yes definitely. I will remove it.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java (hunk 4 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:61]: @ibacher I mistook the Nonnull annotation. I will add it back to it and the base class.

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:61]: Perfectly understandable... My IDE also tells me those checks are useless, but sometimes you have to know better than your tools.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java (hunk 1 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:79]: @ibacher I will change the code accordingly.

**Comment by Medhavi-16** [comment]
@ibacher  I made the said changes. PTAL.

**Comment by theanandankit** [comment]
@Medhavi-16, thanks for this. I think you should add some more tests to cover all the changes.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java [hunk 5]
```diff
@@ -73,6 +74,74 @@ public Encounter get(@Nonnull String uuid) {
...
+		
+		throw new InvalidRequestException("Invalid type of request");
+	}
+	
+	@Override
+	public Encounter delete(@Nonnull String uuid) {
+		
+		if (uuid == null) {
+			throw new InvalidRequestException("Uuid cannot be null.");
+		}
```
Comment (theanandankit) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:131]: Please cover this line with tests.

Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:131]: @theanandankit i will add the test.

**Comment by Medhavi-16** [comment]
@ibacher I added the missing tests and changed things according to your review. PTAL.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java [hunk 6]
```diff
@@ -73,6 +74,74 @@ public Encounter get(@Nonnull String uuid) {
 		return result;
 	}
 	
+	@Override
+	public Encounter create(@Nonnull Encounter encounter) {
+
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:79]:
```suggestion
```

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java [hunk 7]
```diff
@@ -73,6 +74,74 @@ public Encounter get(@Nonnull String uuid) {
...
+		if (result.equals(FhirUtils.OpenmrsEncounterType.ENCOUNTER)) {
+			return super.create(encounter);
+		}
+		
+		if (result.equals(FhirUtils.OpenmrsEncounterType.VISIT)) {
+			return visitService.create(encounter);
+		}
+		
+		throw new InvalidRequestException("Invalid type of request");
+
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:99]:
```suggestion
```

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java (hunk 6 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:79]: Can you please generally remove the blank lines at the start and end of functions?

File: api/src/main/java/org/openmrs/module/fhir2/api/util/FhirUtils.java [hunk 8]
```diff
@@ -100,4 +110,25 @@ public static OperationOutcome createExceptionErrorOperationOutcome(String diagn
 		issue.setDiagnostics(diagnostics);
 		return outcome;
 	}
+	
+	public static Optional<OpenmrsEncounterType> getOpenmrsEncounterType(Encounter encounter) {
+		List<OpenmrsEncounterType> openmrsEncounterTypes = encounter.getType().stream().flatMap(
+		    it -> it.getCoding().stream().filter(coding -> FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem())
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/util/FhirUtils.java:116]: Whitespace:

```suggestion
			it -> it.getCoding().stream().filter(coding -> FhirConstants.ENCOUNTER_TYPE_SYSTEM_URI.equals(coding.getSystem())
```

**Review by ibacher — CHANGES_REQUESTED** [review]
@Medhavi-16 Thanks for this. First of all, I think there's a slight conflict with the other PR I just merged, so if you could fix that, that'd be great.

Also, we need a few more tests here. Yes, they're a little repetitive, but it makes it easier for us to know when and where we're breaking something, so ideally, we'd have some tests in the OMOD for the resource providers in addition to the integration tests and service-layer tests.

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java (hunk 6 above)
Comment (Medhavi-16) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEncounterServiceImpl.java:79]: @ibacher I will remove them.

**Comment by ibacher** [comment]
@Medhavi-16 Do you need any help with this ticket? Thanks again!

**Comment by Medhavi-16** [comment]
> @Medhavi-16 Do you need any help with this ticket? Thanks again!

@ibacher I am waiting for FM2-366 to be resolved and merged. Without that I am facing the error `failed to validate with reason: encounterDatetime: Encounter.datetime.required` in the OMOD tests

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/EncounterFhirResourceProviderIntegrationTest.java [hunk 9]
```diff
@@ -347,4 +358,101 @@ public void shouldReturnCountForEncounterAsXml() throws Exception {
...
+		assertThat(response, isOk());
+		
+		Encounter newEncounter = readResponse(response);
+		
+		assertThat(newEncounter.getId(), equalTo(encounter.getId()));
+		
+	}
+	
+	@Test
+	public void shouldCreateEncounterFromOpenMrsVisitAsJson() throws Exception {
```
Comment (Medhavi-16) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/EncounterFhirResourceProviderIntegrationTest.java:425]: @ibacher This is the integration test for creating Visit type Encounter

Comment (ibacher) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/EncounterFhirResourceProviderIntegrationTest.java:425]: @Medhavi-16: I'd wager the error is caused by [this attribute type](https://github.com/openmrs/openmrs-module-fhir2/blob/16ff04925718bc6b0c718bef0439f48751e81bf9/test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirVisitDaoImplTest_initial_data.xml#L22). Can we try not loading that file and finding another existing visit in the standard data set to work with>?

Comment (Medhavi-16) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/EncounterFhirResourceProviderIntegrationTest.java:425]: @ibacher Thank you! Yes I'll try to use the standard dataset.

**Comment by Medhavi-16** [comment]
@ibacher PTAL.

**Review by ibacher — APPROVED** [review]
@Medhavi-16 Thanks for this! LGTM!
